### PR TITLE
Add docstring and oop array construction method for build_explicit_observed_function 

### PR DESCRIPTION
**Pull request by BenChung**
## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Can we translate this to the standard? https://github.com/SciML/SciMLStyle?tab=readme-ov-file#documentation we should update that to use `##` for the sections but still, the point being that it documents it by arg is a bit easier to read. I don't quite understand how it's written here.

**Comment by BenChung** [comment]
This one is tricky because it's substantially specifying the signature of a method it returns, rather than the method itself. I'm not sure where in the style guide it talks about documenting the argument and return types of returned function values? This is also an issue for SII, too, so it would be worthwhile coming up with a standard.

File: src/systems/diffeqs/odesystem.jl [hunk 1]
```diff
@@ -411,8 +411,32 @@ ODESystem(eq::Equation, args...; kwargs...) = ODESystem([eq], args...; kwargs...
...
+* `p...` unconditionally; note that in the case of `MTKParameters` more than one parameters argument may be present, so it must be splatted
+* `t` if the system is time-dependent; for example `NonlinearSystem` will not have `t`
+For example, a function `g(op, unknowns, p, inputs, t)` will be the in-place function generated if `return_inplace` is true, `ts` is a vector, an array of inputs `inputs` is given, and `params_only` is false for a time-dependent system.
+
+Options not otherwise specified are:
+* `output_type = Array` the type of the array generated by the out-of-place vector-valued function
+* `checkbounds = true` checks bounds if true when destructuring parameters
+* `op = Operator` sets the recursion terminator for the walk done by `vars` to identify the variables that appear in `ts`. See the documentation for `vars` for more detail.
+* `throw = true` if true, throw an error when generating a function for `ts` that reference variables that do not exist
+* `drop_expr` is deprecated.
```
[the remark this answers is not in the document]
Comment (BenChung) [reply src/systems/diffeqs/odesystem.jl:438]: I don't know; it's not actually used for anything in the function so far as I can tell.

That statement comes from Fredrik:
> drop_expr is no longer needed, it was used to control whether or not the Expr was included in the generated function or not, for debugging purposes

[the remark this answers is not in the document]
Comment (ChrisRackauckas) [reply src/systems/diffeqs/odesystem.jl:438]: It's fine to just always require `RuntimeGeneratedFunctions.get_expression` yes, but then if this argument is never documented or used, why mention it in the docstring at all? Just remove it?

[the remark this answers is not in the document]
Comment (BenChung) [reply src/systems/diffeqs/odesystem.jl:438]: I think that when I wrote this I was just trying to document each argument... yeah, it should probably just be deleted. It looks like the only other use is https://github.com/SciML/ModelingToolkit.jl/blob/9cf859fbf05b818e5904aef63d020cbf67e1f199/test/input_output_handling.jl#L144, which is doing exactly nothing.

**Comment by BenChung** [comment]
@ChrisRackauckas What would help a lot is if there's an example/a standard for how to document first-class functions, since that's ultimately what's being described here (albeit sort of in a matrix of the argument configuration to the signature of the output function). I couldn't find it in the style guide you linked, but I might just be missing something.

The proximal problem is that the function (and the functions it returns) behaves very differently depending on the arguments and the system that it's generating an observed function for. It can return one or two functions, and the signatures of the returned functions can have between 0 and 5 arguments depending on a bunch of different properties.

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by ChrisRackauckas** [comment]
Rebase

**Pull request merged**